text/0035-emergency-restarts.md

```diff
@@ -0,0 +1,116 @@
+
+Currently, the post-state of a block at height `h` is defined in the following way:
+- It is the post-state of the block `h-1`
+- with the transforms resulting from executing the deploys contained in block `h` applied
+- with the transforms resulting from executing the `step` function applied, if block `h` contains the `EraEnd` structure.
+
+We need to add one more point to that:
+- with the transforms specified by the upgrade point at height `h` applied (if one exists).
+
+The criteria for validity of a block at height `h` should also take this into account, i.e. a block at height `h` is only valid if its global state hash is the result of applying the above steps to the global state specified at height `h-1`.
+
+Note that this has some desired properties:
+- We can specify any change to the state at any height without breaking the continuity of the chain. The first block after the restart is just like any other regular block, which means that eg. chain synchronization after the restart will work just like before the restart.
+- If the last safe block before the restart was at height `h`, but there were also blocks `h+1...h+n` that were finalized, they will become invalid after an upgrade point is specified at height `h+1`. Since the upgrade point wasn't included in the chainspec before the restart, the global state of the old block `h+1` won't satisfy the validity condition after an upgrade is added and the network is restarted. This means that all the blocks finalized after `h` will be discarded after restarting with the new chainspec and the network will resume operation at height `h+1`.
+
+### Consensus
+
+On the consensus side, it is important for an era change to happen at the boundary where the emergency restart happened. One of the reasons is that the fix for the issue that caused the crash could involve changes to the consensus logic, which could render the state of the last era invalid. This means that the conditions defining a switch block have to change as well: a switch block will be a block that either contains the `EraEnd` structure, or that has height `h-1` if there is an upgrade point specified for height `h`.
+
+### The structure of an upgrade point
+
+Currently, an upgrade point supports defining some new config values. We need it to also be able to specify some transforms that should be applied to the global state. This could be realized in multiple ways:
+
+1. Specifying the modifications as direct key-value pairs to be written to the global state (potentially overwriting the values under existing keys).
+2. Specifying arbitrary WASM to be executed, which would result in the transforms to be applied.
+3. For some special use cases, specifying changes in other ways, eg. support for changes to the validator set to be specified using an `accounts.csv`-like file.
+
+In this CEP, we suggest a combination of options 2 and 3.
```

Comment:
I disagree; option 1 is preferable IMO.

Reply:
I disagree with you Ed, I think specifying an upgrade as a list of `Transforms` (that are internal detail and were supposed to be removed already) is inferior. It will be only understood by the dev team. By using a combination of 2 and 3 we're not tying it to internal details and allowing anyone (that knows Rust) to read the upgrade. Writing `Transforms` "by hand" will also be a difficult task, we have no tooling for that.

Reply:
With option 1 it would be hard to create the set of transforms (we'd need to create an additional tool for that, which we wanted to avoid) and hard to audit by the validators (which are supposed to reach a social consensus on the chainspec, so they should probably be aware of what is contained within). That's why my preference is for a combination of 2 and 3.

Reply:
1) We are not planning to get rid of Transforms afaik, so I'm curious why you think we are.
2) It wouldn't be Transforms; a map of Key, StoredValue is sufficient.
3) We do have tooling to write keyed values to global state; we use it all the time...the TrackingCopy. we would just loop over the map -> tc.write(k,v) applies the necessary state changes, get effects, commit, return the new root hash.

Reply:
Re point 3, I (and Mateusz probably too) mean a tool that would generate such a map of Key → StoredValue from a more high-level description, like "reset the list of validators to this". I'm not sure how easy it would be to just manually write a list of keys to be modified in order to perform consistent changes like that, but I imagine it might be a bit complex and it would be easier to do that internally based on a description that is more human-readable.